File: experimental/e2ee/packet.go
package e2ee

import (
	"bytes"
	"crypto/ed25519"
	"encoding/json"
	"io"
)

// PublicMetadata is an explicit projection, never arbitrary Provider metadata.
// Fields not required for Hub scheduling/permission routing remain encrypted.
type PublicMetadata struct {
	State              string `json:"state,omitempty"`
	Role               string `json:"role,omitempty"`
	RequestID          string `json:"request_id,omitempty"`
	Decision           string `json:"decision,omitempty"`
	Operation          string `json:"operation,omitempty"`
	Outcome            string `json:"outcome,omitempty"`
	CompletionState    string `json:"completion_state,omitempty"`
	ReasonCode         string `json:"reason_code,omitempty"`
	CommandID          string `json:"command_id,omitempty"`
	InterruptSupported bool   `json:"interrupt_supported,omitempty"`
	StopSupported      bool   `json:"stop_supported,omitempty"`
}

type ContentPacket struct {
	Version   int            `json:"version"`
	Public    PublicMetadata `json:"public"`
	Encrypted Envelope       `json:"encrypted"`
}

type packetContent struct {
	Public  PublicMetadata  `json:"public"`
	Payload json.RawMessage `json:"payload"`
}

func (p PublicMetadata) validate(eventType string) error {
	switch eventType {
	case "session.state":
		switch p.State {
		case "starting", "ready", "busy", "waiting_permission", "ended", "error":
		default:
			return ErrInvalid
		}
		if p.Role != "" || p.RequestID != "" || p.Decision != "" {
			return ErrInvalid
		}
	case "session.message":
		if p.Role != "user" && p.Role != "agent" && p.Role != "system" {
			return ErrInvalid
		}
		if p.State != "" || p.RequestID != "" || p.Decision != "" {
			return ErrInvalid
		}
	case "permission.request":
		if !identifier.MatchString(p.RequestID) || p.State != "" || p.Role != "" || p.Decision != "" {
			return ErrInvalid
		}
	case "permission.decision", "permission.respond":
		if !identifier.MatchString(p.RequestID) || p.State != "" || p.Role != "" {
			return ErrInvalid
		}
		if p.Decision != "approve" && p.Decision != "deny" && p.Decision != "expired" {
			return ErrInvalid
		}
	case "session.run.outcome":
		if (p.Operation != "interrupt" && p.Operation != "stop") || p.State != "" || p.Role != "" || p.RequestID != "" || p.Decision != "" {
			return ErrInvalid
		}
		if p.CommandID != "" && !identifier.MatchString(p.CommandID) {
			return ErrInvalid
		}
		switch p.Outcome {
		case "completed", "rejected", "timeout", "outcome_unknown":
		default:
			return ErrInvalid
		}
		if p.CompletionState != "" && p.CompletionState != "ready" && p.CompletionState != "ended" {
			return ErrInvalid
		}
		if p.ReasonCode != "" && !identifier.MatchString(p.ReasonCode) {
			return ErrInvalid
		}
	case "session.run.capabilities":
		if p.State != "" || p.Role != "" || p.RequestID != "" || p.Decision != "" || p.Operation != "" || p.Outcome != "" || p.CompletionState != "" || p.ReasonCode != "" || p.CommandID != "" {
			return ErrInvalid
		}
	default:
		if p != (PublicMetadata{}) {
			return ErrInvalid
		}
	}
	return nil
}

// ProjectPublicMetadata derives, rather than accepts, the public control fields
// from an ordinary domain payload. Everything else remains encrypted.
func ProjectPublicMetadata(eventType string, payload json.RawMessage) (PublicMetadata, error) {
	fields, err := packetFields(payload)
	if err != nil {
		return PublicMetadata{}, err
	}
	var result PublicMetadata
	read := func(name string, target *string) error {
		if fields[name] == nil || json.Unmarshal(fields[name], target) != nil || *target == "" {
			return ErrInvalid
		}
		return nil
	}
	// An optional public field may be absent or JSON null; an empty string is
	// omitted from the projection.
	readOptional := func(name string, target *string) error {
		raw := fields[name]
		if raw == nil || string(raw) == "null" {
			*target = ""
			return nil
		}
		if json.Unmarshal(raw, target) != nil {
			return ErrInvalid
		}
		return nil
	}
	// A capability boolean projection omits false, so absence means unsupported.
	readBool := func(name string, target *bool) error {
		raw := fields[name]
		if raw == nil || string(raw) == "null" {
			*target = false
			return nil
		}
		return json.Unmarshal(raw, target)
	}
	switch eventType {
	case "session.state":
		err = read("state", &result.State)
	case "session.message":
		err = read("role", &result.Role)
	case "permission.request":
		err = read("request_id", &result.RequestID)
	case "permission.decision", "permission.respond":
		err = read("request_id", &result.RequestID)
		if err == nil {
			err = read("decision", &result.Decision)
		}
	case "session.run.outcome":
		err = readOptional("cmd_id", &result.CommandID)
		if err == nil {
			err = read("operation", &result.Operation)
		}
		if err == nil {
			err = read("outcome", &result.Outcome)
		}
		if err == nil {
			err = readOptional("completion_state", &result.CompletionState)
		}
		if err == nil {
			err = readOptional("reason_code", &result.ReasonCode)
		}
	case "session.run.capabilities":
		err = readBool("interrupt_supported", &result.InterruptSupported)
		if err == nil {
			err = readBool("stop_supported", &result.StopSupported)
		}
	}
	if err != nil || result.validate(eventType) != nil {
		return PublicMetadata{}, ErrInvalid
	}
	return result, nil
}

// SealPacket keeps the domain payload unchanged inside the authenticated packet.
// A caller supplies only the control projection approved for the event type.
func SealPacket(ctx Context, key []byte, signer ed25519.PrivateKey, public PublicMetadata, payload json.RawMessage) (ContentPacket, error) {
	projected, err := ProjectPublicMetadata(ctx.Type, payload)
	if err != nil || !sameControlProjection(public, projected) || (public.CommandID != "" && public.CommandID != projected.CommandID) {
		return ContentPacket{}, ErrInvalid
	}
	plaintext, err := json.Marshal(packetContent{public, payload})
	if err != nil {
		return ContentPacket{}, ErrInvalid
	}
	defer clear(plaintext)
	sealed, err := Seal(ctx, key, signer, plaintext)
	if err != nil {
		return ContentPacket{}, err
	}
	return ContentPacket{1, public, sealed}, nil
}

func packetFields(data []byte) (map[string]json.RawMessage, error) {
	decoder := json.NewDecoder(bytes.NewReader(data))
	token, err := decoder.Token()
	if err != nil || token != json.Delim('{') {
		return nil, ErrInvalid
	}
	fields := make(map[string]json.RawMessage)
	for decoder.More() {
		token, err := decoder.Token()
		if err != nil {
			return nil, ErrInvalid
		}
		name, ok := token.(string)
		if !ok {
			return nil, ErrInvalid
		}
		if _, exists := fields[name]; exists {
			return nil, ErrInvalid
		}
		var raw json.RawMessage
		if decoder.Decode(&raw) != nil {
			return nil, ErrInvalid
		}
		fields[name] = raw
	}
	if token, err := decoder.Token(); err != nil || token != json.Delim('}') {
		return nil, ErrInvalid
	}
	if _, err := decoder.Token(); err != io.EOF {
		return nil, ErrInvalid
	}
	return fields, nil
}

// OpenPacket verifies the outside projection against its authenticated copy
// before returning any payload. Hub metadata alone is never endpoint authority.
func OpenPacket(ctx Context, key []byte, signer ed25519.PublicKey, packet ContentPacket) (json.RawMessage, error) {
	if packet.Version != 1 || packet.Public.validate(ctx.Type) != nil {
		return nil, ErrInvalid
	}
	plaintext, err := Open(ctx, key, signer, packet.Encrypted)
	if err != nil {
		return nil, err
	}
	defer clear(plaintext)
	return openPacketPlaintext(ctx.Type, packet.Public, plaintext)
}

func openPacketPlaintext(eventType string, expected PublicMetadata, plaintext []byte) (json.RawMessage, error) {
	fields, err := packetFields(plaintext)
	if err != nil || len(fields) != 2 || fields["public"] == nil || fields["payload"] == nil {
		return nil, ErrInvalid
	}
	public, err := decodePublicMetadata(eventType, fields["public"])
	if err != nil {
		return nil, err
	}
	projected, err := ProjectPublicMetadata(eventType, fields["payload"])
	if err != nil || !sameControlProjection(projected, public) || public != expected || (public.CommandID != "" && public.CommandID != projected.CommandID) {
		return nil, ErrInvalid
	}
	return append(json.RawMessage(nil), fields["payload"]...), nil
}

// sameControlProjection compares a derived projection with a supplied or sealed
// one. command_id is an optional Hub routing hint introduced after endpoints had
// already sealed outcomes, so its absence is not a projection mismatch. The
// authenticated copy and the transport projection still compare exactly.
func sameControlProjection(left, right PublicMetadata) bool {
	left.CommandID, right.CommandID = "", ""
	return left == right
}

func decodePublicMetadata(eventType string, data []byte) (PublicMetadata, error) {
	projection, err := packetFields(data)
	if err != nil {
		return PublicMetadata{}, ErrInvalid
	}
	var public PublicMetadata
	for name, value := range projection {
		var target *string
		switch name {
		case "state":
			target = &public.State
		case "role":
			target = &public.Role
		case "request_id":
			target = &public.RequestID
		case "decision":
			target = &public.Decision
		case "operation":
			target = &public.Operation
		case "outcome":
			target = &public.Outcome
		case "completion_state":
			target = &public.CompletionState
		case "reason_code":
			target = &public.ReasonCode
		case "command_id":
			target = &public.CommandID
		case "interrupt_supported":
			var supported bool
			if json.Unmarshal(value, &supported) != nil {
				return PublicMetadata{}, ErrInvalid
			}
			public.InterruptSupported = supported
			continue
		case "stop_supported":
			var supported bool
			if json.Unmarshal(value, &supported) != nil {
				return PublicMetadata{}, ErrInvalid
			}
			public.StopSupported = supported
			continue
		default:
			return PublicMetadata{}, ErrInvalid
		}
		if json.Unmarshal(value, target) != nil || *target == "" {
			return PublicMetadata{}, ErrInvalid
		}
	}
	if public.validate(eventType) != nil {
		return PublicMetadata{}, ErrInvalid
	}
	return public, nil
}

// DecodeContentPacket is the bounded network decoder. A plain json.Unmarshal
// must not be used at this boundary: it silently accepts duplicate members.
func DecodeContentPacket(eventType string, data []byte) (ContentPacket, error) {
	if len(data) > 48*1024 {
		return ContentPacket{}, ErrInvalid
	}
	fields, err := packetFields(data)
	if err != nil || len(fields) != 3 || fields["version"] == nil || fields["public"] == nil || fields["encrypted"] == nil {
		return ContentPacket{}, ErrInvalid
	}
	var version int
	if json.Unmarshal(fields["version"], &version) != nil || version != 1 {
		return ContentPacket{}, ErrInvalid
	}
	public, err := decodePublicMetadata(eventType, fields["public"])
	if err != nil {
		return ContentPacket{}, err
	}
	envelope, err := DecodeEnvelope(fields["encrypted"])
	if err != nil {
		return ContentPacket{}, err
	}
	return ContentPacket{version, public, envelope}, nil
}

File: client-ts/src/experimental/packet.ts
import { decodeStrictJson } from './strictJson.js'
import { openContent, sealContent, type ContentContext, type ContentEnvelope, type ContentKey } from './e2ee.js'

export interface PublicMetadata {
  state?: string; role?: string; request_id?: string; decision?: string
  operation?: string; outcome?: string; completion_state?: string; reason_code?: string
  interrupt_supported?: boolean; stop_supported?: boolean
}
export interface ContentPacket { version: 1; public: PublicMetadata; encrypted: ContentEnvelope }
function invalid(): never { throw new Error('invalid encrypted packet') }
const runControlTypes = new Set(['session.run.outcome', 'session.run.capabilities'])
function validatePublic(type: string, value: PublicMetadata): void {
  if (!value || typeof value !== 'object' || Array.isArray(value)) invalid()
  const keys = Object.keys(value).sort().join(',')
  switch (type) {
    case 'session.state':
      if (keys !== 'state' || !['starting', 'ready', 'busy', 'waiting_permission', 'ended', 'error'].includes(value.state ?? '')) invalid()
      return
    case 'session.message':
      if (keys !== 'role' || !['user', 'agent', 'system'].includes(value.role ?? '')) invalid()
      return
    case 'permission.request':
      if (keys !== 'request_id' || !validId(value.request_id)) invalid()
      return
    case 'permission.decision':
    case 'permission.respond':
      if (keys !== 'decision,request_id' || !validId(value.request_id) || !['approve', 'deny', 'expired'].includes(value.decision ?? '')) invalid()
      return
    case 'session.run.outcome': {
      // A legacy endpoint sealed an empty projection; either shape is accepted.
      if (keys === '') return
      const allowed = new Set(['operation', 'outcome', 'completion_state', 'reason_code'])
      for (const key of Object.keys(value)) if (!allowed.has(key)) invalid()
      if (!['interrupt', 'stop'].includes(value.operation ?? '')) invalid()
      if (!['completed', 'rejected', 'timeout', 'outcome_unknown'].includes(value.outcome ?? '')) invalid()
      if (value.completion_state !== undefined && !['ready', 'ended'].includes(value.completion_state)) invalid()
      if (value.reason_code !== undefined && !validId(value.reason_code)) invalid()
      return
    }
    case 'session.run.capabilities': {
      const allowed = new Set(['interrupt_supported', 'stop_supported'])
      for (const key of Object.keys(value)) {
        if (!allowed.has(key) || typeof (value as Record<string, unknown>)[key] !== 'boolean') invalid()
      }
      return
    }
    default:
      if (keys !== '') invalid()
  }
}
function validId(value: unknown): boolean { return typeof value === 'string' && /^[A-Za-z0-9_.:/-]{1,128}$/.test(value) }
function normalized(value: PublicMetadata): string {
  return JSON.stringify([
    value.state ?? '', value.role ?? '', value.request_id ?? '', value.decision ?? '',
    value.operation ?? '', value.outcome ?? '', value.completion_state ?? '', value.reason_code ?? '',
    value.interrupt_supported === true, value.stop_supported === true,
  ])
}
export function projectPublicMetadata(type: string, payload: unknown): PublicMetadata {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) invalid()
  const fields = payload as Record<string, unknown>
  let projection: PublicMetadata = {}
  switch (type) {
    case 'session.state': projection = { state: fields.state as string }; break
    case 'session.message': projection = { role: fields.role as string }; break
    case 'permission.request': projection = { request_id: fields.request_id as string }; break
    case 'permission.decision':
    case 'permission.respond': projection = { request_id: fields.request_id as string, decision: fields.decision as string }; break
    case 'session.run.outcome': {
      projection = { operation: fields.operation as string, outcome: fields.outcome as string }
      if (fields.completion_state !== undefined && fields.completion_state !== null) projection.completion_state = fields.completion_state as string
      if (fields.reason_code !== undefined && fields.reason_code !== null) projection.reason_code = fields.reason_code as string
      break
    }
    case 'session.run.capabilities': {
      projectRunControlCapabilities(projection, fields)
      break
    }
  }
  validatePublic(type, projection)
  return projection
}
function projectRunControlCapabilities(projection: PublicMetadata, fields: Record<string, unknown>): void {
  if (fields.interrupt_supported === true) projection.interrupt_supported = true
  if (fields.stop_supported === true) projection.stop_supported = true
}
export async function sealPacket(context: ContentContext, key: ContentKey, signer: CryptoKey, publicMetadata: PublicMetadata, payload: unknown): Promise<ContentPacket> {
  validatePublic(context.type, publicMetadata)
  if (normalized(projectPublicMetadata(context.type, payload)) !== normalized(publicMetadata)) invalid()
  const projection = { ...publicMetadata }
  const plaintext = new TextEncoder().encode(JSON.stringify({ public: projection, payload }))
  return { version: 1, public: projection, encrypted: await sealContent(context, key, signer, plaintext) }
}
// decodeURIComponent uses strict UTF-8 and rejects malformed sequences. Hermes
// supplies TextEncoder but not TextDecoder; keep both runtime paths fail-closed.
export function decodePacketUTF8(bytes: Uint8Array): string {
  if (typeof TextDecoder !== 'undefined') return new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(bytes)
  return decodeURIComponent(Array.from(bytes, byte => '%' + byte.toString(16).padStart(2, '0')).join(''))
}
export async function openPacket(context: ContentContext, key: ContentKey, signer: CryptoKey, packet: ContentPacket): Promise<unknown> {
  try {
    if (!packet || Object.keys(packet).sort().join(',') !== 'encrypted,public,version' || packet.version !== 1) invalid()
    validatePublic(context.type, packet.public)
    const plaintext = await openContent(context, key, signer, packet.encrypted)
    try {
      const content = decodeStrictJson(decodePacketUTF8(plaintext)) as { public: PublicMetadata; payload: unknown }
      if (!content || Object.keys(content).sort().join(',') !== 'payload,public') invalid()
      validatePublic(context.type, content.public)
      if (normalized(content.public) !== normalized(packet.public)) invalid()
      // A legacy endpoint sealed an empty projection for run-control events; its
      // payload still drives the endpoint-owned outcome, so accept the empty
      // public only for those types instead of rejecting valid history.
      const legacyRunControlProjection = runControlTypes.has(context.type) && Object.keys(packet.public).length === 0
      if (!legacyRunControlProjection && normalized(projectPublicMetadata(context.type, content.payload)) !== normalized(content.public)) invalid()
      return content.payload
    } finally { plaintext.fill(0) }
  } catch { return invalid() }
}
